### PR TITLE
Cluster manager task throttling feature [Final PR]

### PR DESCRIPTION
**Pull request by dhwanilpatel**
### Description
Final PR from feature branch (feature/master-task-throttling). Done re base with mainline and raised PR.

Changes were already reviewed and merged in feature branch. Below is list of changes.

* Basic Throttler Framework / Exponential Basic back off policy. (
https://github.com/opensearch-project/OpenSearch/pull/3527 )
* Changes required in Master node to perform throttling.(
https://github.com/opensearch-project/OpenSearch/pull/3882)
* Changes required in Data node to perform retry on throttling.(
https://github.com/opensearch-project/OpenSearch/pull/4204 )
* Provide support for all task type in throttling framework.(
https://github.com/opensearch-project/OpenSearch/pull/4542 )
* Integration Tests (https://github.com/opensearch-project/OpenSearch/pull/4588 )

### Issues Resolved
#479 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/cluster/service/ClusterManagerTaskThrottler.java [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            return taskThrottlingKey;
+        }
+
+        public boolean isThrottlingEnabled() {
+            return throttlingEnabled;
+        }
+    }
+
+    void validateSetting(final Settings settings) {
+        if (minNodeVersionSupplier.get().compareTo(Version.V_2_4_0) < 0) {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/service/ClusterManagerTaskThrottler.java:111]: We might need immediate backport to avoid BWC failures

**Review by Bukhtawar — APPROVED** [review]

**Comment by Bukhtawar** [comment]
Please check for backward compatibility that maybe needed to avoid failures

File: server/licenses/jna-5.5.0.jar.sha1 [hunk 2]
```diff
@@ -1 +1 @@
-0e0845217c4907822403912ad6828d8e0b256208
+0e0845217c4907822403912ad6828d8e0b256208
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply server/licenses/jna-5.5.0.jar.sha1]: Removed. Somehow got added in rebase with mainline.

File: CHANGELOG.md [hunk 3]
```diff
@@ -50,6 +50,13 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Update previous release bwc version to 2.5.0 ([#5003](https://github.com/opensearch-project/OpenSearch/pull/5003))
 - Use getParameterCount instead of getParameterTypes ([#4821](https://github.com/opensearch-project/OpenSearch/pull/4821))
 - Remote shard balancer support for searchable snapshots ([#4870](https://github.com/opensearch-project/OpenSearch/pull/4870))
+- Add Cluster manager task throttling framework ([#4986](https://github.com/opensearch-project/OpenSearch/pull/4986)) ([#479](https://github.com/opensearch-project/OpenSearch/issues/479))
+    - Throttling Exception / New Backoff policy([#3527](https://github.com/opensearch-project/OpenSearch/pull/3527))
+    - Cluster Manager node side change([#3882](https://github.com/opensearch-project/OpenSearch/pull/3882))
+    - Data node side change([#4204](https://github.com/opensearch-project/OpenSearch/pull/4204))
+    - on-boarding of tasks([#4542](https://github.com/opensearch-project/OpenSearch/pull/4542))
+    - Integs ([4588](https://github.com/opensearch-project/OpenSearch/pull/4588))
```
[the remark this answers is not in the document]
Comment (dhwanilpatel) [reply CHANGELOG.md]: Done, Added some details around it. Please let me know if any specific information needs to be added.

File: server/src/main/java/org/opensearch/cluster/service/ClusterManagerTaskThrottler.java (hunk 1 above)
Comment (dhwanilpatel) [reply server/src/main/java/org/opensearch/cluster/service/ClusterManagerTaskThrottler.java:111]: We need to back port it to 2.x branch only right? Will raise PR for back porting to 2.x branch.

**Comment by Bukhtawar** [comment]
```
[org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Snapshot and Restore with repository-s3 using permanent credentials}](https://build.ci.opensearch.org/job/gradle-check/6218/testReport/junit/org.opensearch.repositories.s3/RepositoryS3ClientYamlTestSuiteIT/test__yaml_repository_s3_20_repository_permanent_credentials_Snapshot_and_Restore_with_repository_s3_using_permanent_credentials_/)
[org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Try to create repository with broken endpoint override and named client}](https://build.ci.opensearch.org/job/gradle-check/6218/testReport/junit/org.opensearch.repositories.s3/RepositoryS3ClientYamlTestSuiteIT/test__yaml_repository_s3_20_repository_permanent_credentials_Try_to_create_repository_with_broken_endpoint_override_and_named_client_/)
[org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Snapshot and Restore with repository-s3 using permanent credentials}](https://build.ci.opensearch.org/job/gradle-check/6218/testReport/junit/org.opensearch.repositories.s3/RepositoryS3ClientYamlTestSuiteIT/test__yaml_repository_s3_20_repository_permanent_credentials_Snapshot_and_Restore_with_repository_s3_using_permanent_credentials__2/)
```

**Comment by dhwanilpatel** [comment]
```
[org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Snapshot and Restore with repository-s3 using permanent credentials}](https://build.ci.opensearch.org/job/gradle-check/6233/testReport/junit/org.opensearch.repositories.s3/RepositoryS3ClientYamlTestSuiteIT/test__yaml_repository_s3_20_repository_permanent_credentials_Snapshot_and_Restore_with_repository_s3_using_permanent_credentials_/)
    [org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Register a read-only repository with a non existing client}](https://build.ci.opensearch.org/job/gradle-check/6233/testReport/junit/org.opensearch.repositories.s3/RepositoryS3ClientYamlTestSuiteIT/test__yaml_repository_s3_20_repository_permanent_credentials_Register_a_read_only_repository_with_a_non_existing_client_/)
```

**Comment by dhwanilpatel** [comment]
Gradle check failed with some jenkins issue.

```
java.lang.NullPointerException
	at hudson.FilePath.isAbsolute(FilePath.java:296)
	at hudson.FilePath.resolvePathIfRelative(FilePath.java:281)
	at hudson.FilePath.<init>(FilePath.java:272)
	at hudson.FilePath.child(FilePath.java:1522)
	at hudson.tools.AbstractCommandInstaller.performInstallation(AbstractCommandInstaller.java:86)
	at hudson.tools.InstallerTranslator.getToolHome(InstallerTranslator.java:70)
	at hudson.tools.ToolLocationNodeProperty.getToolHome(ToolLocationNodeProperty.java:108)
	at hudson.tools.ToolInstallation.translateFor(ToolInstallation.java:221)
	at hudson.model.JDK.forNode(JDK.java:149)
	at hudson.model.JDK.forNode(JDK.java:59)
	at org.jenkinsci.plugins.workflow.steps.ToolStep$Execution.run(ToolStep.java:157)
	at org.jenkinsci.plugins.workflow.steps.ToolStep$Execution.run(ToolStep.java:138)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
```

**Comment by reta** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> 
>     * **RESULT:** FAILURE x
> 
>     * **URL:** https://build.ci.opensearch.org/job/gradle-check/6291/
> 
>     * **CommitID:** [f01bc22](https://github.com/opensearch-project/OpenSearch/commit/f01bc22cdf91967cc74a77fbbf26ae82da72d5bf)
>       Please examine the workflow log, locate, and copy-paste the failure below, then iterate to green.
>       Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

@peterzhuamazon it seems like all pipelines are failing with `java.lang.NullPointerException`, could you please take a look? thanks!

**Comment by peterzhuamazon** [comment]
> > ### Gradle Check (Jenkins) Run Completed with:
> > ```
> > * **RESULT:** FAILURE x
> > 
> > * **URL:** https://build.ci.opensearch.org/job/gradle-check/6291/
> > 
> > * **CommitID:** [f01bc22](https://github.com/opensearch-project/OpenSearch/commit/f01bc22cdf91967cc74a77fbbf26ae82da72d5bf)
> >   Please examine the workflow log, locate, and copy-paste the failure below, then iterate to green.
> >   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?
> > ```
> 
> @peterzhuamazon it seems like all pipelines are failing with `java.lang.NullPointerException`, could you please take a look? thanks!

Sorry we fixed it latest Thursday.

**Pull request merged**